### ISSUE TITLE
Sorting nested relation that refer to the same table

### ISSUE THREAD

**Issue by Morinohtar** [yajra/laravel-datatables#2045]
### Issue

Sorting nested relation that refers to the same table.
I tried some solutions that i found here and some others but i'm stuck...

### Code

The function that renders the code is pretty much this:

```php
$foodGroups = FoodGroup::with([
            'parent' => function ($query) use ($locale) {
                $query->select(['id', "name->{$locale} AS nameLocale"]);
            }
        ])
        ->select([
            'food_groups.id',
            'food_groups.parent_id',
            "food_groups.name->{$locale} AS nameLocale",
            'food_groups.published_at',
            'food_groups.created_at',
        ]);

        return DataTables::of($foodGroups)
            ->filterColumn('nameLocale', function ($query, $keyword) use ($locale) {
                $sql = "LOWER(json_unquote(json_extract(`food_groups`.`name`, '$.\"{$locale}\"')))  LIKE ?";

                $query->whereRaw($sql, ["%{$keyword}%"]);
            })
            ->filterColumn('parent.nameLocale', function ($query, $keyword) use ($locale) {
                $query->whereHas('parent', function ($query) use ($keyword, $locale) {
                    $sql = "LOWER(json_unquote(json_extract(`laravel_reserved_0`.`name`, '$.\"{$locale}\"')))  LIKE ?";

                    $query->whereRaw($sql, ["%{$keyword}%"]);
                });
            })
            ->addColumn('checkbox', function ($foodGroup) {
                return view('backend.administrators.pages.food-groups._partials.datatable.checkbox', [
                    'foodGroup' => $foodGroup,
                ]);
            })
            ->editColumn('nameLocale', function ($foodGroup) {
                return view('backend.administrators.pages.food-groups._partials.datatable.name', [
                    'foodGroup' => $foodGroup,
                ]);
            })
            ->editColumn('parent', function ($foodGroup) {
                return view('backend.administrators.pages.food-groups._partials.datatable.parent', [
                    'foodGroup' => $foodGroup,
                ]);
            })
            ->editColumn('published_at', function ($foodGroup) {
                return view('backend.administrators.pages.food-groups._partials.datatable.published-at', [
                    'foodGroup' => $foodGroup,
                ]);
            })
            ->editColumn('created_at', function ($foodGroup) {
                return view('backend.administrators.pages.food-groups._partials.datatable.created-at', [
                    'foodGroup' => $foodGroup,
                ]);
            })
            ->addColumn('actions', function ($foodGroup) {
                return view('backend.administrators.pages.food-groups._partials.datatable.actions', [
                    'foodGroup' => $foodGroup,
                ]);
            })
            ->rawColumns(['checkbox', 'nameLocale', 'parent', 'published_at', 'created_at', 'actions'])
            ->toJson();
```

Error msg:

```
Syntax error or access violation: 1066 Not unique table/alias: 'food_groups' (SQL: select `food_groups`.`id`, `food_groups`.`parent_id`, json_unquote(json_extract(`food_groups`.`name`, '$."pt"')) as `nameLocale`, `food_groups`.`published_at`, `food_groups`.`created_at` from `food_groups` left join `food_groups` on `food_groups`.`parent_id` = `food_groups`.`id` order by `food_groups`.`nameLocale` asc limit 25 offset 0)
```

For the search filters i managed to resolved it using the `laravel_reserved_0` that is automatically generated by your code as an `alias` for the table. But for the sorting i havent figure it out how to "fix" this.
Maybe you could create an `alias`, like you did for the filters, when making the `join` code, so the tables won't collide when doing the ordering.

### System details

- PHP Version: 7.3.x
- Laravel Version: 5.8.x
- Laravel-Datatables Version: 9.x

Thanks,
Luís

**Comment by Morinohtar** [comment]
I just did a tiny test. I changed the function `joinEagerLoadedColumn` in `EloquentDataTable.php` and i got it to work. 
What I did regards the `BelongsTo` relations. And this will break the others in some way. But like i said, this was just to see if it would work, giving an alias to the joined table.

I'm sure you can get the idea of what i did and build a decent fix for this issue. Assuming there isnt a better solution or something that already exists and i can use.

Changed code:

```php
protected function joinEagerLoadedColumn($relation, $relationColumn)
    {
        $table     = '';
        $deletedAt = false;
        $lastQuery = $this->query;
        foreach (explode('.', $relation) as $index => $eachRelation) { // changed
            $model = $lastQuery->getRelation($eachRelation);
            switch (true) {
                case $model instanceof BelongsToMany:
                    $pivot   = $model->getTable();
                    $pivotPK = $model->getExistenceCompareKey();
                    $pivotFK = $model->getQualifiedParentKeyName();
                    $this->performJoin($pivot, $pivotPK, $pivotFK);

                    $related = $model->getRelated();
                    $table   = $related->getTable();
                    $tablePK = $related->getForeignKey();
                    $foreign = $pivot . '.' . $tablePK;
                    $other   = $related->getQualifiedKeyName();

                    $lastQuery->addSelect($table . '.' . $relationColumn);
                    $this->performJoin($table, $foreign, $other);

                    break;

                case $model instanceof HasOneOrMany:
                    $table     = $model->getRelated()->getTable();
                    $foreign   = $model->getQualifiedForeignKeyName();
                    $other     = $model->getQualifiedParentKeyName();
                    $deletedAt = $this->checkSoftDeletesOnModel($model->getRelated());
                    break;

                case $model instanceof BelongsTo:
                    $table     = $model->getRelated()->getTable();
                    $tableAlias = "nested_{$index}_{$table}"; // changed
                    $alias     = $model->getRelated()->getTable() . " AS $tableAlias"; // changed
                    $foreign   = $model->getQualifiedForeignKeyName();
                    $other     = $model->getQualifiedOwnerKeyName();
                    $owner     = $tableAlias . '.' . $model->getOwnerKeyName(); // changed
                    $deletedAt = $this->checkSoftDeletesOnModel($model->getRelated());
                    break;

                default:
                    throw new Exception('Relation ' . get_class($model) . ' is not yet supported.');
            }

            $this->performJoin($alias, $foreign, $owner, $deletedAt); // changed
            $lastQuery = $model->getQuery();
        }

        return $tableAlias . '.' . $relationColumn; // changed
    }
```

**Comment by yajra** [comment]
@Morinohtar yes, self join is not yet supported by the package. Can you please submit a PR so that we can review further and might work with a solution. Thanks!

**Comment by yajra** [comment]
Fixed via #2051.
